File: puzzle5.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <vector>
#include <iterator>
using namespace std;

std::string part2(const std::vector<int> &input)
{
    std::vector<int> l(input);
    int cnt = 0;
    int idx = 0;
    while ( idx < l.size() ) {
        int jmp = l[idx];
        if (jmp >= 3) {
            l[idx]--;
        } else {
            l[idx]++;
        }
        idx += jmp;
        cnt++;
    }
    std::stringstream sstr;
    sstr << cnt;
    return sstr.str(); 
}

int main()
{
    std::cout << part2(std::vector<int>( std::istream_iterator<int>(std::cin), std::istream_iterator<int>() )) << std::endl;
    return 0;
}

File: puzzle5.py
import sys

def part1(numbers):
    cnt = 0
    l = list(numbers)
    idx = 0
    while idx < len(l):
        jmp = l[idx]
        l[idx] += 1
        idx += jmp
        cnt += 1
    return cnt

def part2(numbers):
    cnt = 0
    l = list(numbers)
    idx = 0
    while idx < len(l):
        jmp = l[idx]
        if jmp >= 3:
            l[idx] -= 1
        else:
            l[idx] += 1
        idx += jmp
        cnt += 1
    return cnt

if __name__ == "__main__":
    lines = [int(line.strip()) for line in sys.stdin.readlines()]
    print part1(lines)
    print part2(lines)
